File: hw2_stack/hw2.java
import java.util.*;
import java.lang.*;
import java.io.*;

public class hw2 {
	public static String readFile(String fileName) throws IOException {
                BufferedReader br = new BufferedReader(new FileReader(fileName));
                try {
                        StringBuilder sb = new StringBuilder();
                        String line = br.readLine();

                        while(line != null) {
                                sb.append(line);
                                sb.append("\n");
                                line = br.readLine();
                        }
                        return sb.toString();
                }
                finally {
                        br.close();
                }
        }
	public static void hw2(String input, String output)throws IOException {
		//read input file
		String in = new String();
                in = readFile(input);
		
		//output file	
                PrintWriter pw = new PrintWriter(new FileWriter(output));
               	
		//create stack of Objects
		Stack<String> stack = new Stack<String>();
		
		//Strings for testing
		String strx = new String();
		String stry = new String();
		
		//int for testing
		int intx;
		int inty;
	
		//array of lines of files
                String[] file_str = in.split("\n"); // each index of array is one line of the file
		
		//parse file_string
		for (int i = 0; i < file_str.length; i++) {
			String[] str = file_str[i].split(" "); // first index of array is the action, second is the variable
			//quit
			if (str[0].equals("quit")) {
				while(!stack.empty()) {	
					pw.write(stack.pop()+"\n");
				}		
			}
			//push
			else if (str[0].equals("push")) {	
				stack.push(str[1]);
			}
			//pop
			else if (str[0].equals("pop")) {
				if(!stack.empty()) stack.pop();
				else stack.push(":error:");
			}
			//boolean
			else if (str[0].equals(":true:") || str[0].equals(":false:")) {
				stack.push(str[0]);
			}
			//error
			else if (str[0].equals(":error:")) {
				stack.push(str[0]);
			}
			//add
			else if (str[0].equals("add")) {	
				stry = stack.peek();
				if (stack.empty() || stack.size() == 1) stack.push(":error:");
				else if(!stry.equals(":error:") && !stry.equals(":false:") && !stry.equals(":true:")) {
					stry = stack.pop();
					strx = stack.peek();
					if(!strx.equals(":error:") && !strx.equals(":false:") && !strx.equals(":true:")) {
						strx = stack.pop();
						inty = Integer.parseInt(stry);
						intx = Integer.parseInt(strx);
						int addition = intx + inty;
						StringBuilder sb = new StringBuilder();
						sb.append("");
						sb.append(addition);
						String add = sb.toString();
						stack.push(add); 
					}
					else {
						stack.push(stry);
						stack.push(":error:");
					}	
				}
			}
			//sub
			else if (str[0].equals("sub")) {
				stry = stack.peek();
				if (stack.empty() || stack.size() == 1) stack.push(":error:");
				else if(!stry.equals(":error:") && !stry.equals(":false:") && !stry.equals(":true:")) {
					stry = stack.pop();
					strx = stack.peek();
					if(!strx.equals(":error:") && !strx.equals(":false:") && !strx.equals(":true:")) {
						strx = stack.pop();
						inty = Integer.parseInt(stry);
						intx = Integer.parseInt(strx);
						int subtraction = intx - inty;
						StringBuilder sb = new StringBuilder();
						sb.append("");
						sb.append(subtraction);
						String sub = sb.toString();
						stack.push(sub);
					}
					else {
						stack.push(stry);
						stack.push(":error:");
					}
				} 
			}
			//mul
			else if (str[0].equals("mul")) {
                                stry = stack.peek();
				if (stack.empty() || stack.size() == 1) stack.push(":error:");
                                else if(!stry.equals(":error:") && !stry.equals(":false:") && !stry.equals(":true:")) {
                                        stry = stack.pop();
					strx = stack.peek();
                                        if(!strx.equals(":error:") && !strx.equals(":false:") && !strx.equals(":true:")) {
						strx = stack.pop();
                                                inty = Integer.parseInt(stry);
                                                intx = Integer.parseInt(strx);
                                                int multiplication = intx * inty;
                                                StringBuilder sb = new StringBuilder();
                                                sb.append("");
                                                sb.append(multiplication);
                                                String mul = sb.toString();
                                                stack.push(mul);
                                        }
					else {
						stack.push(stry);
						stack.push(":error:");
					}
                                }
                        }
			//div
			else if (str[0].equals("div")) {
                                stry = stack.peek();
                                if (stack.empty() || stack.size() == 1) stack.push(":error:");
                                else if(!stry.equals(":error:") && !stry.equals(":false:") && !stry.equals(":true:") && !stry.equals("0")) {
                                        stry = stack.pop();
					strx = stack.peek();
                                        if(!strx.equals(":error:") && !strx.equals(":false:") && !strx.equals(":true:")) {
						strx = stack.pop();
                                                inty = Integer.parseInt(stry);
                                                intx = Integer.parseInt(strx);
                                                int division = intx / inty;
                                                StringBuilder sb = new StringBuilder();
                                                sb.append("");
                                                sb.append(division);
                                                String div = sb.toString();
                                                stack.push(div);
                                        }
					else {
						stack.push(stry);
						stack.push(":error:");
					}
                                }
                        }
	        	//rem
	        	else if (str[0].equals("rem")) {
				stry = stack.peek();
				if (stack.empty() || stack.size() == 1) stack.push(":error:");
                                else if(!stry.equals(":error:") && !stry.equals(":false:") && !stry.equals(":true:") && !stry.equals("0")) {
                                        stry = stack.pop();
					strx = stack.peek();
                                        if(!strx.equals(":error:") && !strx.equals(":false:") && !strx.equals(":true:")) {
						strx = stack.pop();
                                                inty = Integer.parseInt(stry);
                                                intx = Integer.parseInt(strx);
                                                int remainder = intx % inty;
                                                StringBuilder sb = new StringBuilder();
                                                sb.append("");
                                                sb.append(remainder);
                                                String rem = sb.toString();
                                                stack.push(rem);
                                        }
					else {
						stack.push(stry);
						stack.push(":error:");
					}
                                }
			}
			//neg
			else if (str[0].equals("neg")) {
				stry = stack.peek();
				if(stack.empty()) stack.push(":error:");
				else if(!stry.equals(":error:") && !stry.equals(":false:") && !stry.equals(":true:")) {
					stry = stack.pop();
					inty = Integer.parseInt(stry);
                                        int negative = inty * -1;
                                        StringBuilder sb = new StringBuilder();
                                        sb.append("");
                                        sb.append(negative);
                                        String neg = sb.toString();
                                        stack.push(neg) ;
				}
				else stack.push(":error:");
			}
			//swap
			else if (str[0].equals("swap")) {
				if(stack.empty() || stack.size() == 1) stack.push(":error:");
				else {
					stry = stack.pop();
					strx = stack.pop();
					stack.push(stry);
					stack.push(strx);
				}
			}
		}
		pw.close();
	}
}

File: hw2_stack/hw2.py
import sys

class Stack:
	def __init__(self):
		self.items = []
	def isEmpty(self):
		return self.items == []
	def push(self, item):
		self.items.append(item)
	def pop(self):
		return self.items.pop()
	def size(self):
		return len(self.items)

def hw2(input_file, output_file):
	f = open(input_file, 'r')
	read = f.read()
	inp = read.splitlines()
	out = open(output_file, 'r+')
	stack = Stack()
	for s in inp:
		string = s.split()
		if string[0] == "quit":
			length = stack.size()
			while length != 0:
				x = stack.pop()
				length = stack.size()
				out.write(str(x)+"\n")
		elif string[0] == "pop":
			length = stack.size()
			if length == 0:
				stack.push(":error:")
			else:
				stack.pop()
		elif string[0] == ":error:":
			stack.push(":error:")
		elif string[0] == ":false:":
			stack.push(":error:")
		elif string[0] == ":true:":
			stack.push(":true:")
		elif string[0] == "push":
			stack.push(int(string[1]))
		elif string[0] == "add":
			length = stack.size()
			y = stack.pop()
			if length < 2:
				stack.push(y)
				stack.push(":error:")
			elif y != ":error:" and y != ":false:" and y != ":true:":
				x = stack.pop()
				if x != ":error:" and x != ":false:" and x != ":true:":
					add = x + y
					stack.push(int(add))
				else:
					stack.push(x)
					stack.push(y)
					stack.push(":error:")
		elif string[0] == "sub":
			length = stack.size()
			y = stack.pop()
			if length < 2:
				stack.push(y)
				stack.push(":error:")
			elif y != ":error:" and y != ":false:" and y != ":true:":
				x = stack.pop()
				if x != ":error:" and x != ":false:" and x != ":true:":
					sub = x - y
					stack.push(int(sub))
				else:
					stack.push(x)
					stack.push(y)
					stack.push(":error:")
		elif string[0] == "mul":
			length = stack.size()
			y = stack.pop()
			if length < 2:
				stack.push(y)
				stack.push(":error:")
			elif y != ":error:" and y != ":false:" and y != ":true:":
				x = stack.pop()
				if x != ":error:" and x != ":false:" and x != ":true:":
					mul = x * y
					stack.push(int(mul))
				else:
					stack.push(x)
					stack.push(y)
					stack.push(":error:")
		elif string[0] == "div":
			length = stack.size()
			y = stack.pop()
			if length < 2:
				stack.push(y)
				stack.push(":error:")
			elif y != ":error:" and y != ":false:" and y != ":true:":
				x = stack.pop()
				if x != ":error:" and x != ":false:" and x != ":true:":
					div = x / y
					stack.push(int(div))
				else:
					stack.push(x)
					stack.push(y)
					stack.push(":error:")
		elif string[0] == "rem":
			length = stack.size()
			y = stack.pop()
			if length < 2:
				stack.push(y)
				stack.push(":error:")
			elif y != ":error:" and y != ":false:" and y != ":true:":
				x = stack.pop()
				if x != ":error:" and x != ":false:" and x != ":true:":
					rem = x % y
					stack.push(int(rem))
				else:
					stack.push(x)
					stack.push(y)
					stack.push(":error:")
		elif string[0] == "neg":
			length = stack.size()
			y = stack.pop()
			if length < 0:
				stack.push(y)
				stack.push(":error:")
			elif y != ":error:" and y != ":false:" and y != ":true:":
				neg = y * -1
				stack.push(int(neg))
			else:
				stack.push(y)
				stack.push(":error:")
		elif string[0] == "swap":
			length = stack.size()
			if length < 2:
				stack.push(":error:")
			else:
				y = stack.pop()
				x = stack.pop()
				stack.push(y)
				stack.push(x)				
	out.close()
	f.close()
